subsys/bluetooth/host/direction.c

```diff
@@ -429,6 +459,140 @@ static int hci_df_set_conn_cte_tx_param(struct bt_conn *conn, uint8_t cte_types,
+			 */
+			cp->slot_durations = BT_HCI_LE_ANTENNA_SWITCHING_SLOT_2US;
+			cp->switch_pattern_len = ARRAY_SIZE(df_dummy_switch_pattern);
+			ant_ids = &df_dummy_switch_pattern[0];
+		}
+
+		dest_ant_ids = net_buf_add(buf, params->num_ant_ids);
+		memcpy(dest_ant_ids, params->ant_ids, params->num_ant_ids);
+	}
+}
+
+static int hci_df_set_conn_cte_rx_enable(struct bt_conn *conn, bool enable,
+					 const struct bt_df_conn_cte_rx_param *params)
+{
+	struct bt_hci_rp_le_set_conn_cte_rx_params *rp;
+	struct bt_hci_cmd_state_set state;
+	struct net_buf *buf, *rsp;
+	int err;
+
+	if (enable) {
+		err = validate_cte_rx_common_params(params->cte_type, params->slot_durations,
+						    params->num_ant_ids, params->ant_ids);
+		if (err != 0) {
+			return err;
+		}
+	}
+
+	/* If CTE Rx is enabled, command parameters total length must include
+	 * antenna ids, so command size if extended by num_and_ids.
+	 */
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_CONN_CTE_RX_PARAMS,
+				(sizeof(struct bt_hci_rp_le_set_conn_cte_rx_params) +
+				 (enable ? params->num_ant_ids : 0)));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	prepare_conn_cte_rx_enable_cmd_params(buf, conn, params, enable);
+
+	bt_hci_cmd_state_set_init(buf, &state, conn->flags, BT_CONN_CTE_RX_ENABLED, enable);
+
+	err = bt_hci_cmd_send_sync(BT_HCI_OP_LE_SET_CONN_CTE_RX_PARAMS, buf, &rsp);
+	if (err) {
+		return err;
+	}
+
+	rp = (void *)rsp->data;
+	if (conn->handle != sys_le16_to_cpu(rp->handle)) {
+		err = -EIO;
+	} else {
+		conn->cte_type = (enable ? params->cte_type : 0);
+	}
+
+	net_buf_unref(rsp);
+
+	return err;
+}
+
+int hci_df_prepare_connection_iq_report(struct net_buf *buf,
+					 struct bt_df_conn_iq_samples_report *report,
+					 struct bt_conn **conn_to_report)
+{
+	struct bt_hci_evt_le_connection_iq_report *evt;
+	struct bt_conn *conn;
+
+	if (buf->len < sizeof(*evt)) {
+		BT_ERR("Unexpected end of buffer");
+		return -EINVAL;
+	}
+
+	evt = net_buf_pull_mem(buf, sizeof(*evt));
+
+	conn = bt_conn_lookup_handle(sys_le16_to_cpu(evt->conn_handle));
+
+	if (!conn) {
+		BT_ERR("Unknown conn handle 0x%04X for iq samples report",
+		       sys_le16_to_cpu(evt->conn_handle));
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(conn->flags, BT_CONN_CTE_RX_ENABLED)) {
+		BT_ERR("Received conn CTE report when CTE receive disabled");
+		return -EINVAL;
+	}
+
+	if (!(conn->cte_type & BIT(evt->cte_type))) {
+		BT_DBG("CTE filtered out by cte_type: %u", evt->cte_type);
+		return -EINVAL;
+	}
+
+	report->chan_idx = evt->data_chan_idx;
+	report->rx_phy = evt->rx_phy;
+	report->chan_idx = evt->data_chan_idx;
+	report->rssi = evt->rssi;
+	report->rssi_ant_id = evt->rssi_ant_id;
+	report->cte_type = BIT(evt->cte_type);
+	report->packet_status = evt->packet_status;
+	report->slot_durations = evt->slot_durations;
+	report->conn_evt_counter = sys_le16_to_cpu(evt->conn_evt_counter);
+	report->sample_count = evt->sample_count;
+	report->sample = &evt->sample[0];
```

Comment:
Given that both report->sample and evt->sample is an array of samples then we probably should do
```suggestion
	report->sample = evt->sample;
```
right?